Sprint 14

ihatelags opened the pull request:
feat: added Booking controller
feat: added Comments
feat: added JPA repositories

ihatelags committed:
Sprint 14
feat: added Booking controller
feat: added Comments
feat: added JPA repositories

ihatelags committed:
Sprint 14
feat: added Booking controller
feat: added Comments
feat: added JPA repositories

ihatelags committed:
fixed codestyle errors

ihatelags committed:
fixed codestyle errors

utausheva commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,123 @@
+DROP TABLE IF EXISTS users, request, items, booking, comments CASCADE;
+
+CREATE TABLE IF NOT EXISTS users
```
довольно тяжело читается такая схема) Давай одно поле в одну строку писать?)

ihatelags replied:
я писал это в нормальном виде. Даже не знаю как оно в такой вид превратилось

utausheva commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -2,16 +2,40 @@
     private String email;
+
+    @Override
+    public boolean equals(Object o) {
```
методы equals и hashcode действительно нужны тебе? Расскажи где и зачем?

ihatelags replied:
equals нужен был для одной из валидаций. Я потом переписал, забыл убрать.

utausheva commented on src/main/java/ru/practicum/shareit/booking/dto/BookingOutputDto.java:
```diff
@@ -0,0 +1,23 @@
+public class BookingOutputDto {
+    private long id;
+    private UserDto booker;
+    private ItemDto item;
```
Одно дто не может ссылаться на другое дто. Вообще дто не может быть сложных типов данных. Здесь должен быть внутренний класс Item, вот объяснение наставника:
BookingDto.Item и ItemDto - разное назначение.
То что все поля из BookingDto.Item есть в ItemDto - это повезло. но логически они разные и могут быть разные.
Т.е. не должно быть причины добавлять поле в ItemDto , только потому что это потребовалось в BookingDto.Item

Подправь во всех дто всех сущностей.

ihatelags replied:
поправил

utausheva reviewed (commented):
Привет!
Всё отлично, но надо подправить дто)

ihatelags committed:
фиксы по результам ревью:
Убрал лишние equals и hashcode в моделях
В bookingDto исправил тип полей
schema.sql - исключил из автореформата

utausheva commented on src/main/java/ru/practicum/shareit/booking/dto/BookingOutputDto.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+public class BookingOutputDto {
+    private long id;
+    private User booker;
```
а проблема не решилась, эти классы всё еще должны быть внутренними)) Логика та же: если тебе понадобится, например, добавить поле для вывода в класс юзера, не будешь ведь ты в модель добавлять это поле)

utausheva commented on src/main/java/ru/practicum/shareit/item/comment/CommentDto.java:
```diff
@@ -0,0 +1,17 @@
+    private long id;
+    private String text;
+    private String authorName;
+    private ItemDto item;
```
и здесь

utausheva commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -19,6 +19,5 @@ public class ItemDto {
     private Boolean available;
-    private User owner;
-    private ItemRequest request;
+    private List<CommentDto> comments;
```
и CommentDto тоже сделать внутренним

utausheva commented on src/main/java/ru/practicum/shareit/item/dto/ItemDtoWithBooking.java:
```diff
@@ -0,0 +1,26 @@
+    private String description;
+    @NotNull(groups = OnCreate.class)
+    private Boolean available;
+    private List<CommentDto> comments;
```
и здесь

utausheva commented on src/main/java/ru/practicum/shareit/item/dto/ItemDtoWithBooking.java:
```diff
@@ -0,0 +1,26 @@
+    @NotNull(groups = OnCreate.class)
+    private Boolean available;
+    private List<CommentDto> comments;
+    private BookingOutputDtoForItem lastBooking;
```
если BookingOutputDtoForItem используется только в этом классе, то можно оставить как есть, если нет, то тоже сделать внутренним

ihatelags replied:
Да, используется только здесь

utausheva reviewed (commented):
дто надо добить))

ihatelags committed:
фиксы по результам ревью:
сделал внутренние классы в ДТОшках

utausheva reviewed (approved):
Теперь отлично!
Можно мерджить))
